LFS-1233: As a questionnaire designer, I can specify descriptions for predefined answer options

veronikaslc opened the pull request.

Commit:
LFS-1233: As a questionnaire designer, I can specify descriptions for predefined answer options

sashaandjic commented:
Can't build
![image](https://user-images.githubusercontent.com/23243244/125708886-079961f3-996c-4bcb-893e-c223efc87980.png)

Commit:
LFS-1233: As a questionnaire designer, I can specify descriptions for predefined answer options
Added missing file

sashaandjic commented:
I can define descriptions for predefined answer options, however I do not see them in the form at all. When I refreshed my form I got this in the console:
![image](https://user-images.githubusercontent.com/23243244/125901160-6e819584-255c-49a9-9374-bfe88996b55a.png)

Commit:
LFS-1233: As a questionnaire designer, I can specify descriptions for predefined answer options
Added description to the Form

sdumitriu commented on modules/commons/src/main/frontend/src/components/ComposedIcon.jsx:
```diff
@@ -0,0 +1,78 @@
+//
```
This was copied from an MIT-licensed source. I don't think we can just relicense it as Apache. It's better to just leave the MIT license on top, along with a "copied from... " comment, and add an exception in the rat configuration.

sdumitriu reviewed (commented):
The description doesn't look too good as it is, I would make it look more like the question description (small and grey).

sashaandjic commented:
Smaller fonts would look better. Also, dark gray color may improve it too.
![image](https://user-images.githubusercontent.com/23243244/126232744-b9b2892a-22ff-4a42-bbf7-1eea14e7e4cb.png)

Commit:
LFS-1233: As a questionnaire designer, I can specify descriptions for predefined answer options
Addressed codereview comment

sashaandjic commented:
I do not like the red fonts. Can we try gray instead?
![image](https://user-images.githubusercontent.com/23243244/126270956-b6f81731-c04f-41dc-a472-4ce4856951f0.png)

Commit:
LFS-1233: As a questionnaire designer, I can specify descriptions for predefined answer options
Fixed color for text

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -608,6 +608,26 @@ const questionnaireStyle = theme => ({
+    },
+    selectionDescription: {
+      fontSize: "0.75rem",
+      color: theme.palette.grey["500"],
```
The theme color used should be `theme.palette.text.secondary`.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -571,6 +573,9 @@ function ResponseChild(props) {
                   label: classes.inputLabel
                 }}
               />
+              <br/>
+              <MDEditor.Markdown className={classes.selectionDescription} source={description} />
```
To ensure consistency with the [section description](https://github.com/data-team-uhn/cards/blob/ed1019d0857eac1ea903a3362f77d8d137b317aa/modules/data-entry/src/main/frontend/src/questionnaire/Section.jsx#L80), we should wrap this in a caption typography and inherit its fontsize:
```suggestion
              <Typography variant="caption" component="div" color="textSecondaty">
                <MDEditor.Markdown className={classes.selectionDescription} source={description} />
              </Typography>
```

marta- replied:
Going a bit further, I'm not a big fan of seeing a component with "Editor" in its name being used to **display** text.

I suggest creating a separate component `FormattedText` to be used as follows:
```
<FormattedText variant={any valid Typography varian; optional} color={any valid Typography color; optional}>
  {the text to render}
</FormattedText>
```
For example, in the case of our descriptions displayed in the form:
```
<FormattedText variant="caption" color="textSecondary">
  {description}
</FormattedText>
```

Internally, the component would render:

```
<Typography variant={variant} component="div" color={color}
    <MDEditor.Markdown className={classes.formattedText} source={children} />
</Typography>
```
and would incorporate the styling for `classes.formattedText` to ensure the `fontSize` is `inherit`ed.

veronikaslc replied:
Why do you want to pass description text as a node and not as an another prop?

marta- replied:
I find the usage of the component more logical with the text inside than as a prop, the way it is for Typography for example.

I'm also hoping have a broader use for this component, for example for formatted messages to the user, e.g. like this (completely made up, with excessive formatting just to serve as an example):

```
<Alert severity="error">
  <FormattedText>
      You cannot create more than **eight** forms for this subject. If you wish to add more information of this type:
      * delete an existing form before creating a new one, *or*
      * modify an existing form
  </FormattedText>
</Alert>
```
Using markdown to specify formatting in such messages will be much easier for the dev than using html markup and styling. But when we have long texts on multiple lines, they're easier to read and write as a child of the component than as a prop.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -608,6 +608,26 @@ const questionnaireStyle = theme => ({
+      display: "block"
+    },
+    selectionDescription: {
+      fontSize: "0.75rem",
```
Once this is wrapped in a `Typography` component (see other comment), `fontSize` should be switched to  
```suggestion
      fontSize: "inherit",
```

marta- commented on modules/commons/src/main/frontend/src/components/ComposedIcon.jsx:
```diff
@@ -0,0 +1,84 @@
+        color={color}
+        style={{
+          fontSize: sizesMap[size].extraSize,
+          color: disabled ? 'rgba(0, 0, 0, 0.35)' : null,
```
It's strange that only the bottom icon is disabled. Regardless, it should use a theme color (`theme.palette.text.disabled`) instead of a hardcoded color.

marta- replied:
> It's strange that only the bottom icon is disabled. Regardless, it should use a theme color (`theme.palette.text.disabled`) instead of a hardcoded color.

This hasn't been addressed yet. ^

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/AnswerOptions.jsx:
```diff
@@ -323,6 +357,47 @@ let AnswerOptions = (props) => {
+        disableBackdropClick
+        disableEscapeKeyDown
+        open={Boolean(descriptionAnchorEl)}
+        anchorEl={descriptionAnchorEl}
+        onClose={handleClose}
+        anchorOrigin={{
+          vertical: 'bottom',
+          horizontal: 'center',
+        }}
+        transformOrigin={{
+          vertical: 'top',
+          horizontal: 'center',
+        }}
+        className={classes.descriptionPopover}
+      >
+        { descriptionIndex != null &&
+          <StyledMarkdownTextField
+            objectKey="description"
+            data={options[descriptionIndex]}
+            onChange={setDescription}
+          />
+        }
+        <Button
+          variant='contained'
+          color='default'
+          onClick={handleClose}
+          className={classes.descriptionPopoverButton}
+        >
+          Cancel
+        </Button>
+        <Button
+          variant='contained'
+          color='primary'
+          onClick={onDescriptionPopoverClose}
+          disabled={!Boolean(description)}
+          className={classes.descriptionPopoverButton}
+        >
+          Done
+        </Button>
+    </Popover>
```
This Popover looks quite messy, and the horizontal layout inside it wastes a lot of space.

Suggested changes (see also mockup):
* Switch to vertical layout ("Description" above the editor as a title, not to its left)
* "Description" should use the `h6` typography variant instead of `subtitle1`
* There should be 24px (`theme.spacing(3)`) between the title and the editor box
* The title should specify whose description this is (`Description for "<label of the option>"`)
* There should be 24px (`theme spacing(3)`) between the editor box and the action buttons 
* There should be 8px (`theme.spacing(1)`) between the 2 buttons
* There should not be a vertical scrollbar by default

![image](https://user-images.githubusercontent.com/651980/126366526-4abd2ad0-aed7-4108-91df-698185f16ad8.png)

marta- replied:
Additionally, adding a description should be possible for the special answer options as well (na, noneOfTheAbove).

sashaandjic commented:
The latest look before refactoring

**Edit mode:**
![image](https://user-images.githubusercontent.com/23243244/126400452-c33e58f7-5d81-4b8e-85c3-bb23f56034b6.png)

**View mode:**
![image](https://user-images.githubusercontent.com/23243244/126400510-916abb28-d8ab-4ab0-aa07-cd5d1b37efb2.png)

Commit:
LFS-1233: As a questionnaire designer, I can specify descriptions for predefined answer options
Added FormattedText component to display markdown text in a view mode
Added MarkdownText component with help button

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/NewFormDialog.jsx:
```diff
@@ -37,6 +36,7 @@ import NewItemButton from "../components/NewItemButton.jsx";
 import ResponsiveDialog from "../components/ResponsiveDialog"; // commons
 import QuestionnaireStyle from "../questionnaire/QuestionnaireStyle.jsx";
 import { fetchWithReLogin, GlobalLoginContext } from "../login/loginDialogue.js";
+import FormattedText from "../questionnaire/FormattedText.jsx";
```
`FormattedText` should be in `components`. Also, the file itself seems to be missing from this last commit.

Commit:
LFS-1233: As a questionnaire designer, I can specify descriptions for predefined answer options
Added description to the special answer options

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/AnswerOptions.jsx:
```diff
@@ -285,12 +335,26 @@ let AnswerOptions = (props) => {
+                              MainIcon={NotesIcon}
+                              ExtraIcon={!value.description ? AddCircleIcon : EditIcon}/>
+                          </IconButton>
+                        </Tooltip>
```
The same markup is used in 2 places, with slightly different callbacks. Let's move it to a function and give it the `onClick` callback as a parameter.

marta- replied:
Also,all the "description" icons should be aligned, and the special option switches should be aligned with the delete buttons. Right now the tools for special options are shifted to the left.

![image](https://user-images.githubusercontent.com/651980/126536401-16b82936-f7a4-4b69-a8c6-65c57a3c6fd5.png)

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/MarkdownText.jsx:
```diff
@@ -0,0 +1,59 @@
+
+// Markdown Text Input field used by Edit dialog component
+let MarkdownText = (props) => {
+  let { text, onChange, classes } = props;
```
I'd rename the `text` prop to `value`. This is technically an input, and `value` is consistent with other input controls.

marta- commented on modules/commons/src/main/frontend/src/components/ComposedIcon.jsx:
```diff
@@ -0,0 +1,84 @@
+        lineHeight: '0px',
+        cursor: 'pointer',
+        position: 'absolute',
+        textShadow: '0.75px 0px 0.5px #FFF, 0px 0.75px 0.5px #FFF, -0.75px 0px 0.5px #FFF, 0px -0.75px 0.5px #FFF',
```
Looks like I forgot to comment on this in the last review: `#FFF` won't look nice on dark themes. Let's replace it with `theme.palette.background.default`.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -571,6 +573,11 @@ function ResponseChild(props) {
                   label: classes.inputLabel
                 }}
               />
+              <br/>
```
This newline doesn't appear to be necessary, due to the block display of the description.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/AnswerOptions.jsx:
```diff
@@ -323,6 +387,49 @@ let AnswerOptions = (props) => {
+        className={classes.descriptionPopover}
+      >
+        <Typography variant="h6" className={classes.descriptionPopoverTitle} >
+          {"Description for " + descriptionLabel}
```
The label itself should appear in quotation marks on the UI. This should do it (not tested):
```suggestion
          {`Description for "${descriptionLabel}"`}
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -571,6 +573,11 @@ function ResponseChild(props) {
                 }}
               />
+              <br/>
+              <FormattedText className={classes.selectionDescription} variant="caption" color="textSecondary">
```
**Regression**: The styling specified by selectionDescription is ignored because the `FormattedText` component is not aware of the `classes`.

![image](https://user-images.githubusercontent.com/651980/126544957-020ada21-a4b2-42e3-9523-e94eed2a0852.png)

See fix suggestion in the comment on the `FormattedText` component.

marta- commented on modules/commons/src/main/frontend/src/components/FormattedText.jsx:
```diff
@@ -0,0 +1,55 @@
+}));
+
+let FormattedText = (props) => {
+  let { children, variant, display, color } = props;
+  const classes = useStyles();
+
+  return (
+    <Typography variant={variant} display={display} component="div" color={color} >
+      <MDEditor.Markdown className={classes.markdown} source={children} />
+    </Typography>
```
Here's a suggestion to fix the "discarded styling" bug from the previous comment (tested briefly). `typographyProps` will include `classes` and `className` set when instantiating the `FormattedText` component. This also simplifies the implementation by sending all props that we might set on `FormattedText` directly to `Typography`. The use of the  `markdown` className is restricted only to the internal `MDEditor` component.

```suggestion
  let { children, ...typographyProps } = props;
  const mdClasses = useStyles();

  return (
    <Typography component="div" {...typographyProps} >
      <MDEditor.Markdown classes={mdClasses} className={mdClasses.markdown} source={children} />
    </Typography>
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/AnswerOptions.jsx:
```diff
@@ -234,6 +262,27 @@ let AnswerOptions = (props) => {
+    setIsSpecialOption(false);
+  }
+
+  let onDescriptionPopoverClose = () => {
+    // update corresponding option description
+    if (isSpecialOption) {
+      specialOptionsInfo[descriptionIndex].setter({ ...specialOptionsInfo[descriptionIndex].data, "description": description});
+    } else {
+      setOptions(oldValue => {
+        var value = oldValue.slice();
+        value[descriptionIndex].description = description;
+        return value;
+      });
+    }
+    handleClose();
```
**Bug** (not sure if the responsible code is here or somewhere else):
* lfs runmode -> Administration -> Questionnaire -> Patient information
* Edit the Life status question
* Click "Description" for the first option -> Description popover opens, 'Done" button is disabled (correct)
* Type in a description (e.g. bla) and click Done (now enabled)
* Click Description for the second option -> Description popover opens, **'Done" button is enabled (wrong)**
* Click Done without typing anything in -> **Icon turns into "Notes + Edit" (wrong)**
* Click Description for the second option again -> **observe that the description from the first option was copied in (wrong)**

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -608,6 +605,27 @@ const questionnaireStyle = theme => ({
+    descriptionPopoverTitle: {
+      marginBottom: theme.spacing(3),
+    },
+    selectionChild: {
+      display: "block"
+    },
```
**Bug:** This CSS rule causes layout issues for nested input (e.g. Vocabulary question with maxAnswers=1 and with predefined options, see screenshot). It appears to be unnecessary because the description already has block display which ensures its correct positioning, and should probably be removed.

![image (45)](https://user-images.githubusercontent.com/651980/126554437-4b0622f6-945f-43d9-99dd-34452e73b2ab.png)

sashaandjic commented:
A very small complaint about scaling down zoom.
Here is the definition:
![image](https://user-images.githubusercontent.com/23243244/126559569-5ced27de-cc6c-4917-8527-ac311ac1c333.png)

100% zoom
![image](https://user-images.githubusercontent.com/23243244/126559642-4d13ac8b-d1c7-45bc-b744-c19cc620e882.png)

80%
![image](https://user-images.githubusercontent.com/23243244/126559660-e4e2a7d6-7f23-4834-8557-feb97faf0f88.png)

sashaandjic commented:
Can we align `Caucasian ` and `*Also known as European`?
![image](https://user-images.githubusercontent.com/23243244/126560787-791e82e5-26fa-4745-a509-39df1d471eb6.png)

marta- commented:
> Can we align `Caucasian ` and `*Also known as European`?
> ![image](https://user-images.githubusercontent.com/23243244/126560787-791e82e5-26fa-4745-a509-39df1d471eb6.png)

Yep, it's reported above, it's a small regression after the last refactoring and there's a fix suggestion.

marta- commented:
> A very small complaint about scaling down zoom.

The basic markdown syntax doesn't support image sizing. This particular editor supports HTML syntax out of the box, so you can resize the image by using HTML syntax instead of markdown syntax to display it:

```
<img src="URL of the image" alt="my image" style="width:200px;"/>
```

However, allowing just any plain HTML is a security hole and we should probably disable it. @sdumitriu do you want HTML disabled before merging?

Since the markdown editor is available only to admins for now, it's not too bad to require using images that are appropriately sized. I would say user-friendly image sizing is currently out of scope for now and should be reported as a separate task.

sashaandjic commented:
The problem that I reported is between link "Wikipedia" and text that follows it. On 80% zoom there is a large hole between them. For some reason removing the image that is positioned right before "Wikipedia" link solves the problem.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Question.jsx:
```diff
@@ -38,7 +38,7 @@ function Question (props) {
         title={text}
         titleTypographyProps={{ variant: 'h6' }}
-        subheader={<MDEditor.Markdown className={classes.markdown} source={description} />}
+        subheader={<FormattedText>{description}</FormattedText>}
         subheaderTypographyProps={{ variant: 'caption', component: "div"}}
```
**Regression** (discovered by @sashaandjic): the font size of question descriptions is no longer consistent with the other descriptions because `subheaderTypographyProps` are not applied to the inner Typography of `FormattedText`. To do: move `variant: caption` from `subheaderTypographyProps` to `FormattedText`.

```
        subheader={<FormattedText variant="caption">{description}</FormattedText>}
        subheaderTypographyProps={{component: "div"}}
```

marta- commented:
> The problem that I reported is between link "Wikipedia" and text that follows it. On 80% zoom there is a large hole between them. For some reason removing the image that is positioned right before "Wikipedia" link solves the problem.

I see, I misunderstood the issue. The problem is that the image is displayed inline per source code, and as a large image it will cause gaps in the text. There's nothing we can fix on our side, we just need to make sure our descriptions are properly formatted.

Commit:
LFS-1233: As a questionnaire designer, I can specify descriptions for predefined answer options
Addressed codereview comments

sashaandjic reviewed (approved):
tested - 3 previously reported problems are now fixed

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Question.jsx:
```diff
@@ -38,8 +38,8 @@ function Question (props) {
-        subheader={<MDEditor.Markdown className={classes.markdown} source={description} />}
-        subheaderTypographyProps={{ variant: 'caption', component: "div"}}
+        subheader={<FormattedText variant="caption">{description}</FormattedText>}
+        subheaderTypographyProps={{ component: "div"}}
```
```suggestion
        subheaderTypographyProps={{ component: "div" }}
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -597,17 +594,44 @@ const questionnaireStyle = theme => ({
+    descriptionPopover: {
+      "& .MuiPopover-paper" : {
+        padding: theme.spacing(3),
+        width: theme.spacing(87),
+        height: theme.spacing(40),
```
Do these really need to be specified?

veronikaslc replied:
We do

marta- commented on modules/commons/src/main/frontend/src/components/FormattedText.jsx:
```diff
@@ -0,0 +1,55 @@
+
+FormattedText.propTypes = {
+  children: PropTypes.string,
+  variant: PropTypes.string,
+  color: PropTypes.string,
+  display: PropTypes.string
+};
+
+FormattedText.defaultProps = {
+  variant: 'body1',
+  color: 'initial',
+  display: 'initial'
```
These lines can now be removedbfor simplicity, since we're defaulting to whatever is specified by `Typography`.
```suggestion
```

marta- reviewed (commented):
Looks great, one last **usability bug**: I cannot completely remove a description once entered.

Steps to reproduce:
* Patient information -> Gender
* Enter a description for Male, click done, save the question
* Edit the question again and delete the description -> the Done button becomes disabled once the description is empty, and the removal cannot be saved.

Commit:
LFS-1233: As a questionnaire designer, I can specify descriptions for predefined answer options
Addressed codereview comments; Enabled Done button

sashaandjic reviewed (approved):
tested

sdumitriu reviewed (approved):
Looks good.

@marta- One minor comment about this PR: should clicking on the description select the option (checkbox or radio)? Right now, clicking on the description text doesn't but clicking on the option label does.

Some other observations:
- javascript injection seems to not be easy to do, which is a good thing!
- adding a checkbox in a list currently has bad alignment, but that should be fixed upstream

![image](https://user-images.githubusercontent.com/88663/126880173-642a00fc-073e-4978-aaef-c841f89a3aa9.png)

sashaandjic commented:
> Looks good.
> 
> @marta- One minor comment about this PR: should clicking on the description select the option (checkbox or radio)? Right now, clicking on the description text doesn't but clicking on the option label does.

Logged in [JIRA]( https://phenotips.atlassian.net/browse/CARDS-1275)

> Some other observations:
> 
> * javascript injection seems to not be easy to do, which is a good thing!
> * adding a checkbox in a list currently has bad alignment, but that should be fixed upstream
> 
> ![image](https://user-images.githubusercontent.com/88663/126880173-642a00fc-073e-4978-aaef-c841f89a3aa9.png)

Logged in [JIRA](https://phenotips.atlassian.net/browse/CARDS-1274)
